dbInterface.js

```diff
@@ -0,0 +1,86 @@
+'use strict';
+
+const { Pool } = require('pg');
+
+class DbInterface {
```

Comment:
You probably have started noticing this file is not going to scale. It has too many responsibilities and it'll grow to be a God class.

In nodejs (and in general) you should aim to have the least responsibilities as possible.

I recommend the following:
1. Make this an abstraction of the db
2. Follow the Repository Pattern or write services

### 1. Make this an abstraction of the db 
Node people like the convenience of `index.js` files. Let's try it to see how it works.

Create a `app/db/index.js` file with the following:
```js
// app/db/index.js
const { Pool } = require('pg');

const pool = new Pool({ connectionString: 'postgresql://localhost/user_auth' });

// publishing my interface and keeping stuff private
module.exports = {
  query: (text, params, callback) => pool.query(text, params, callback)
}
```

With this you can now go to your node console and use it:
```js
> db = require('./app/db')
> db.query('select 1', (err, res) => console.log(res.rows))
[ { '?column?': 1 } ]
```

You have successfully abstracted both the connection details and the db interface. This means that if you change the database or the npm package you just have to change this file.

I'm going to go a bit further, I'm not a fan of the current `query` implementation for the PG package. But I saw you can pass an object [like so](https://node-postgres.com/features/queries#query-config-object):
```js
const query = {
  text: 'INSERT INTO users(name, email) VALUES($1, $2)',
  values: ['brianc', 'brian.m.carlson@gmail.com'],
}
```

This means we could probably use a [tagged template](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#Tagged_templates), which I'm clearly a fan of:
```js
const sql = (strings, ...values) => {
  const text = strings.reduce((query, current, idx) => `${query}$${idx}${current}`);
  return { text, values };
}

// with this you can just
username = 'david.marchante@rubiconmd.com';
sql`select * from users where username = ${username};`

// These would be the entry args (you can see we'd just pass through 'values')
// { strings: [ 'select * from users where username = ', ';' ],
//   values:  [ 'david.marchante@rubiconmd.com' ] }
//
// And this the reconstructed query
// 'select * from users where username = $1;'

// So now you could:
const user = await db.query(sql`select * from users where username = ${username};`);
```

### 2. Follow the Repository Pattern or write services
I think you should write **services**, but what you were trying to do was basically a repository pattern.

If you want to do a repository pattern, follow the following structure:
```
repositories/
  userRepository.js   
  sessionRepository.js
  // and so on
```

```js
class userRepository {  // could probably extend a common interface
  constructor(db) {}  // always pass an instance of db to invert the dependency

  find(id): User {}
  create(user: User): User {}
  update
  remove
  // and so on
}
```

I don't like the repository pattern very much because it's a responsibility magnet. Instead I like to use services and query objects:

```js
services/
  createUser.js
queries/
  latestOffers.js
```

### Extra:
I see you made a `rowExists` method, but it's a bit expensive BE side right now. 
When you query for the existence of something you don't need to bring any data with you, just a boolean. I recommend using `SELECT EXISTS(subquery);`:
```js
module.exports = {
  query: (text, params, callback)  => pool.query(text, params, callback),
  exists: async (sqlObject) => {
    sqlObject.text = `select exists(${sqlObject.text});`;
    const result = await pool.query(sqlObject);
    return result.rows[0].exists;
  }
};

// in other file
await db.exists(sql`select * from user_auth.auth where username = ${'david.marchante@rubiconmd.com'}`);
// true
```

Reply:
I only had a little bit of time and got through understanding and coding up section 1. This is such a difference though, the db abstraction and tag are fantastic. Did you just then think of making that sql tag or is this a normal pattern? It's super useful!

Also, for file structure I nested the tag in the db dir, does that feel right? Or is it super granular and that should even be `app/sql/index.js`?

Reply:
It's the first time I go this low level with node. The tagged template just felt natural after refactoring the db object.

Refactoring the db object to expose just a minimal interface is indeed a common pattern though. It's called _Dependency Inversion_ (the one Dependency Injection is based on) and it basically means that instead of depending on low level modules you depend on abstractions (so you _invert_ the dependency). This means you can swap objects (and mock them) even if you change libraries.

As for where to put the `sql` function, in `db/` is fine. I wouldn't go too crazy with granularity.

Reply:
Commits for these comments are pushed up. I went with the services and queries approach. Let me know if that looks like what you were thinking. I've now fully removed that original DB god class! This `sql` tag function is so nice...

Reply:
Hahaha, I'm glad you liked it!

Reply:
I didn't talk, but maybe I should've, about the Active Record pattern —the one we use in Rails—.

The Active Record pattern is a favorite for a lot of ORMs and it makes the model responsible of its own data layer abstraction as well. So you have things like (Rails' ActiveRecord ORM):
```rb
user = User.find_by(name: 'Lee')
user.name = 'Lee 🧛🏻‍♂️'
user.save
```

Separating the business logic from the data layer would be the Data Mapper pattern, which may not look as pretty, but is more easily optimizable.

There's no silver bullet and in the end what matters is that you know when to use what.